src/execution_plan/optimizations/apply_join.c

```diff
@@ -13,7 +13,10 @@
 
 // Tests to see if given filter can act as a join condition.
 static bool _applicableFilter(const FT_FilterNode *f) {
-	return (f->t == FT_N_PRED && f->pred.op == OP_EQUAL);
+	return (f->t == FT_N_PRED && f->pred.op == OP_EQUAL &&
+			// TODO allowing AR_ExpNodes containing graph entities currently causes memory errors
+			f->pred.lhs->type == AR_EXP_OPERAND &&
```

Comment:
Hmm, you're right, the logic here is not quite correct.

The case we want to block here is AR_ExpNodes that represent pure `Node` entities:
`MATCH (a), (b) WHERE a = b`
Since that causes issues with the caching logic in the ValueHashJoin code.